packages/contracts/contracts/plugin/psp/PluginSetupProcessor.sol

```diff
@@ -0,0 +1,703 @@
+        }
+
+        PluginRepo.Version memory currentVersion = _params.pluginSetupRepo.getVersion(
+            _params.currentVersionTag
+        );
+
+        PluginRepo.Version memory newVersion = _params.pluginSetupRepo.getVersion(
+            _params.newVersionTag
+        );
+
+        if (currentVersion.pluginSetup == newVersion.pluginSetup) {
+            // revert plugin setups can't be the same(pointless)
+        }
+
+        bytes32 pluginId = _getPluginId(_dao, _params.setupPayload.plugin);
+
+        PluginInformation storage pluginInformation = states[pluginId];
+
+        bytes32 setupId = _getSetupId(
+            PluginSetupRef(_params.currentVersionTag, _params.pluginSetupRepo),
+            bytes32(0),
+            hHash(_params.setupPayload.currentHelpers),
+            bytes32(0),
+            bytes("")
+        );
+
+        // The following check implicitly confirms that plugin
+        // is currently installed. Otherwise, currentSetupId wouldn't be set.
+        if (pluginInformation.currentSetupId != setupId) {
+            revert InvalidSetupId({
+                currentSetupId: pluginInformation.currentSetupId,
+                setupId: setupId
+            });
+        }
+
+        // Prepare the update.
+        (initData, preparedDependency) = PluginSetup(newVersion.pluginSetup).prepareUpdate(
+            _dao,
+            _params.currentVersionTag.build,
+            _params.setupPayload
+        );
+
+        bytes32 newSetupId = _getSetupId(
+            PluginSetupRef(_params.newVersionTag, _params.pluginSetupRepo),
+            pHash(preparedDependency.permissions),
+            hHash(preparedDependency.helpers),
+            aHash(preparedDependency.actions),
+            initData
+        );
+
+        // Only allow to prepare if setupId has not been prepared before.
+        // Note that the following check ensures that the same setupId can be prepared
+        // once again if the plugin was uninstalled and then installed..
+        if (pluginInformation.blockNumber < pluginInformation.setupIds[newSetupId]) {
+            revert SetupAlreadyPrepared(setupId);
+        }
+
+        pluginInformation.setupIds[newSetupId] = block.number;
+
+        emit UpdatePrepared({
+            sender: msg.sender,
+            dao: _dao,
+            setupId: newSetupId,
+            pluginSetupRepo: _params.pluginSetupRepo,
+            versionTag: _params.newVersionTag,
+            setupPayload: _params.setupPayload,
+            preparedDependency: preparedDependency,
+            initData: initData
+        });
+
+        return (initData, preparedDependency);
+    }
+
+    /// @notice Applies the permissions of a prepared update of an UUPS upgradeable contract to a DAO.
+    /// @param _dao The address of the updating DAO.
+    /// @param _params The struct containing the parameters for the `applyInstallation` function.
+    function applyUpdate(address _dao, ApplyUpdate calldata _params)
```

Comment:
we need a check here that the new setupid isn't an uninstall. currently, you can pass the parameters for a prepared uninstallation into `applyUpdate`. But I am not sure what harm it could do

Reply:
Amazing catch. Fixed in https://github.com/aragon/core/pull/164/commits/4963fa34d5ae4f231204ece3a080504003d557f2 hopefully.